docs/enhancements/status-reporting.md

```diff
@@ -0,0 +1,158 @@
+# Status Reporting
+
+<!-- markdown-toc start - Don't edit this section. Run M-x markdown-toc-refresh-toc -->
+**Table of Contents**
+
+- [Status Reporting](#status-reporting)
+    - [Summary](#summary)
+    - [Objective](#objective)
+        - [Goals](#goals)
+        - [Non-goals](#non-goals)
+    - [Proposal](#proposal)
+        - [For Sonobuoy Aggregator](#for-sonobuoy-aggregator)
+        - [For `sonobuoy status`](#for-sonobuoy-status)
+        - [For Scanner](#for-scanner)
+        - [Heartbeats (optional/future)](#heartbeats-optionalfuture)
+        - [User Stories](#user-stories)
+    - [Unresolved Questions](#unresolved-questions)
+
+<!-- markdown-toc end -->
+
+## Summary
+
+Sonobuoy -- both `sonobuoy run` and through Scanner -- has really lackluster UX.
+At present the tests will be kicked off and no information will be reported
+until the run either succeeds or times out. This is not ideal. We would like to
+have a concept of test status that can be reported, in near-real time, back to
+the user running the service. Both the CLI and Scanner should be able to display
+this progress in a way that makes sense for their individual UX.
+
+## Objective
+
+`sonobuoy run` and Scanner should report the status of an ongoing Sonobuoy test suite.
+This status should be secure so only the intended audience can see it.
+
+### Goals
+
+* Create a command `sonobuoy status` that shows the results of an ongoing run.
+* Create an API on Aggregator so Scanner can report status instead of the
+  existing spinner
+* (optional) Workers can report their status to the aggregator
+
+### Non-goals
+
+* Actually updating Scanner is outside of our scope, though we should make sure
+  the API is useful to that team
+* Complex metrics and monitoring. The first pass should be simple.
+
+## Proposal
+
+### For Sonobuoy Aggregator
+
+In addition to the HTTPS server running for the plugin workers, a second HTTPS
+server and accompanying k8s `service` will be run for the purpose of
+communicating status back to the client. This service will be `NodePort` or
+`LoadBalanacer` rather than the `ClusterIP` of the aggregation service.
+
+To secure this, the client will provide a server TLS cert/key to serve HTTPS
+with, and a CACertificate to validate the client's peer certificates with.
+
+This secondary HTTPS server will report status initially based on the percentage
+of plugins that have reported results. Later, the workers will be able to report
+heartbeats or more granular status.
+
+The API for this status will look like:
+
+`GET /status`
+
+```json
+{
+  "status": "running",
+  "percent_complete": 0.5
```

Comment:
I don't know how you can compute percent complete unless we plumb some form of ginkgo hook.

Reply:
The percent is just % of nodes which have marked themselves as complete